wasm/wasm_source/src/vp_user.rs

```diff
@@ -158,35 +136,200 @@ fn validate_tx(
+                                        | BelowThreshold
+                                )
+                        {
+                            address == owner && **valid_sig
+                        } else if
+                        // Bonding and unbonding may affect validator sets
+                        matches!(
+                            pre,
+                            Consensus | BelowCapacity | BelowThreshold
+                        ) && matches!(
+                            post,
+                            Consensus | BelowCapacity | BelowThreshold
+                        ) {
+                            true
+                        } else {
+                            // Unknown state changes are not allowed
+                            false
+                        }
+                    }
+                    (None, Some(_post)) => {
+                        // Becoming a validator must be authorized
+                        **valid_sig
+                    }
+                    (Some(_pre), None) => {
+                        // Clearing of old epoched data
+                        true
+                    }
+                    _ => false,
+                }
+            }
+            None => false,
+        };
+
+        VpResult::Ok(
+            is_valid_state
+                || storage::is_validator_state_epoched_meta_key(key)
+                || storage::is_consensus_validator_set_key(key)
+                || storage::is_below_capacity_validator_set_key(key),
+        )
+    };
+
+    let is_valid_reward_claim = || {
+        if let Some(bond_id) = storage::is_last_pos_reward_claim_epoch_key(key)
+        {
+            // Claims for this address must be signed
+            return &bond_id.source != owner || **valid_sig;
+        }
+        false
+    };
+
+    let is_valid_redelegation = || {
+        if storage::is_validator_redelegations_key(key) {
+            return true;
+        }
+        if let Some(delegator) = storage::is_delegator_redelegations_key(key) {
+            // Redelegations for this address must be signed
+            return delegator != owner || **valid_sig;
+        }
+        if let Some(bond_id) = storage::is_rewards_counter_key(key) {
+            // Redelegations auto-claim rewards
+            return &bond_id.source != owner || **valid_sig;
+        }
+        false
+    };
+
+    let is_valid_become_validator = || {
+        if storage::is_validator_addresses_key(key)
+            || storage::is_consensus_keys_key(key)
+            || storage::is_validator_eth_cold_key_key(key).is_some()
+            || storage::is_validator_eth_hot_key_key(key).is_some()
+            || storage::is_validator_max_commission_rate_change_key(key)
+                .is_some()
+            || storage::is_validator_address_raw_hash_key(key).is_some()
+        {
+            // A signature is required to become validator
+            return **valid_sig;
+        }
+        false
+    };
```

Comment:
should this look at the data in these keys from `pre` and `post` and be valid if the `pre`s are `None` and `post`s are `Some`?

Reply:
I think we don’t need to go that deep in here. The detailed PoS validation should happen in PoS VP and this VP should only be concerned with authorization